### PR TITLE
Debug commands: Fix level transitions

### LINKED ISSUE #2765: Dungeon items appearing in town
do changelevel 0
image

[single_33.zip](https://github.com/diasurgical/devilutionX/files/7086585/single_33.zip)

Must be dItem/Items/whatever not being cleared

### PR DESCRIPTION
**Pull request by obligaron**
Could fix #2765 😉 

Background:
We start the level transition with `StartNewLvl`.
But `StartNewLvl` only `PostMessage` a level transition to the message queue.
That means that the level transition happens a lot later.
But we change the `currlevel` direclty after `StartNewLvl`.
This is a problem for the save game logic.

When you transit from level to another following code is executed:
```
		IncProgress();
		if (!gbIsMultiplayer) {
			SaveLevel();
		} else {
			DeltaSaveLevel();
		}
		IncProgress();
		FreeGameMem();
		currlevel--;
		leveltype = gnLevelTypeTbl[currlevel];
		assert(myPlayer.plrlevel == currlevel);
		IncProgress();
		LoadGameLevel(false, ENTRY_PREV);
		IncProgress();
```

First is the current level saved and `SaveLevel` uses `currlevel`.
That means we set store items, monster etc. in the wrong level.

Additional fixes:

- Not available maps are can't be accessed by `map` command
- If you leave a map the normal way, it should port you back to the expected position and not (0, 0). 😉

### CODE REVIEW

File: Source/interfac.cpp [hunk 1]
```diff
@@ -358,6 +359,7 @@ void ShowProgress(interface_mode uMsg)
 		}
 		IncProgress();
 		FreeGameMem();
+		setlevel = false;
```
[the remark this answers is not in the document]
Comment (obligaron) [reply Source/interfac.cpp]: `setlevel = false;` is only set in the case for `WM_DIABTWARPUP` and `WM_DIABNEXTLVL`.
But I tested again and I think `WM_DIABTWARPUP` is wrong, cause it is the command for going up from dungeon to town.
I replaced it with correct `WM_DIABTOWNWARP`.

[the remark this answers is not in the document]
Comment (AJenbo) [reply Source/interfac.cpp]: Oh your right, it's just GitHub having some issues regarding indentation lately:

image

File: Source/debug.cpp [hunk 2]
```diff
@@ -184,22 +197,23 @@ std::string DebugCmdLoadMap(const string_view parameter)
...
 	if (level < 1)
 		return fmt::format("Map id must be 1 or higher", level);
-	if (setlevel && myPlayer.plrlevel == level)
-		return fmt::format("I did nothing but fulfilled your wish. You are already at level {}.", level);
+	if (setlevel && setlvlnum == level)
+		return fmt::format("I did nothing but fulfilled your wish. You are already at mapid {}.", level);
 
 	for (auto &quest : Quests) {
-		if (level != quest._qslvl)
+		if (level != quest._qslvl || quest._qactive == quest_state::QUEST_NOTAVAIL)
```
Comment (AJenbo) [inline Source/debug.cpp:204]: Won't this result in it reporting maps for inactive quests to be reported as unknown?

I's the a good reason we should not allow the player to jump to a level from an inactive quesets when debugging?

Comment (qndel) [reply Source/debug.cpp:204]: inactive quest maps have no exit so you end up at 0,0 if you leave them manually

Comment (AJenbo) [reply Source/debug.cpp:204]: That isn't really an issue when debugging, just `changelevel 0` and your good.

Comment (obligaron) [reply Source/debug.cpp]: Personally I would disable the map if the quest is not active.
But I don't want to argue too much about a debug command, so I removed the commit. 😉

**Pull request merged**